src/BenchmarkDotNet/Diagnosers/EventPipeProfiler.cs

```diff
@@ -0,0 +1,241 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Analysers;
+using BenchmarkDotNet.Engines;
+using BenchmarkDotNet.Environments;
+using BenchmarkDotNet.Exporters;
+using BenchmarkDotNet.Extensions;
+using BenchmarkDotNet.Helpers;
+using BenchmarkDotNet.Jobs;
+using BenchmarkDotNet.Loggers;
+using BenchmarkDotNet.Reports;
+using BenchmarkDotNet.Running;
+using BenchmarkDotNet.Validators;
+using JetBrains.Annotations;
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Symbols;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Etlx;
+using Microsoft.Diagnostics.Tracing.Stacks;
+using Microsoft.Diagnostics.Tracing.Stacks.Formats;
+
+namespace BenchmarkDotNet.Diagnosers
+{
+    public class EventPipeProfiler : IProfiler
+    {
+        // parameterless constructor required by DiagnosersLoader to support creating this profiler via console line args
+        public EventPipeProfiler() { }
+
+        /// <summary>
+        /// Creates new instance of EventPipeProfiler
+        /// </summary>
+        /// <param name="profile">A named pre-defined set of provider configurations that allows common tracing scenarios to be specified succinctly.</param>
+        /// <param name="providers">A list of EventPipe providers to be enabled.</param>
+        public EventPipeProfiler(EventPipeProfile? profile = null, IReadOnlyCollection<EventPipeProvider> providers = null)
```

Comment:
this ctor is very complex. It would be better to have two ctors:

```cs
public EventPipeProfiler(IReadOnlyCollection<EventPipeProvider> providers)

public EventPipeProfiler(EventPipeProfile profile) : this(Map(profile))

private static IReadOnlyCollection<EventPipeProvider> Map(EventPipeProfile profile) ...

Reply:
Your suggestion look good but it doesn't have the same function as my constructor.

You can call my constructor with parameters:
```c#
EventPipeProfiler(
    EventPipeProfile.CpuSampling, 
    new[] {
        new EventPipeProvider("System.Buffers.ArrayPoolEventSource", EventLevel.Informational, Int64.MaxValue),
    });
```
It give you all providers from `EventPipeProfile.CpuSampling` and add the `System.Buffers.ArrayPoolEventSource` provider. 
In this way, you can also override the provider from the profile. e.g.:
the `EventPipeProfile.CpuSampling` profile adds:
```c#
new[]
{
    new EventPipeProvider("Microsoft-DotNETCore-SampleProfiler", EventLevel.Informational),
    new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Informational, (long) ClrTraceEventParser.Keywords.Default)
}
```
and you can override for example EventLevel or change keywords, using the following code
```
EventPipeProfiler(
    EventPipeProfile.CpuSampling, 
    new[] {
        new EventPipeProvider("Microsoft-DotNETCore-SampleProfiler", EventLevel.Verbous),
    });
```
Currently it works in the same way like `dotnet trace`. You can fine more information here https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-trace#options-1

But if you want I will change it. Please let me know.

Reply:
thanks for the explanation, now it makes sense to me!